envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -55,10 +71,11 @@ class RBACFilterFactory(
                 return@forEach
             }
 
-            val clients = resolveClients(incomingEndpoint, incomingPermissions.roles)
-            val principals = clients.flatMap { mapClientToPrincipals(it, snapshot) }
+            val clientsWithSelectors = resolveClientsWithSelectors(incomingEndpoint, incomingPermissions.roles)
+            val principals = clientsWithSelectors.flatMap { mapClientWithSelectorToPrincipals(it, snapshot) }.toSet()
```

Comment:
I wonder if `toSet()` is needed at the end of this line.

Reply:
This was added so that header principals are not duplicated, it was discovered in this version: `797653240c554ed2282b7c0770f6b72b5b49cb03`

Reply:
Thanks for info. I think we should definitely remove header principals at some point, but I would do it in separate PR. 
I see now that `mapClientWithSelectorToPrincipals` can generate one Principal max. So there is no point in doing `toSet()` but also no cost :) So I have no preference here, it is up to you.